File: leetcode/Python/0059-m-spiral-matrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0] * n for i in range(n)]
        nums = [i for i in range(1, n*n+1)]
        seen = [[False] * n for i in range(n)]
        dr = [0,1,0,-1]
        dc = [1,0,-1,0]
        cr, cc, r, c, di = 0, 0, 0, 0, 0
        for i in range(n*n):
            res[r][c] = nums[i]
            seen[r][c] = True
            cr, cc = r + dr[di], c + dc[di]
            if 0 <= cr < n and 0 <= cc < n and seen[cr][cc] == False:
                r, c = cr, cc
            else:
                di = (di + 1) % 4
                r, c = r + dr[di], c + dc[di]
        return res

File: leetcode/C++/0059-m-spiral-matrix-ii.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        vector<int> steps = {n , n-1};
        vector<vector<int>> matrix(n, vector<int>(n));
        int dIdx = 0, sIdx = 0;
        int r = 0, c = -1;
        int num = 1;
        while (steps[sIdx] > 0) {
            for (int i = 0; i < steps[sIdx]; i++) {
                r += directions[dIdx][0], c += directions[dIdx][1];
                matrix[r][c] = num;
                num++;
            }
            steps[sIdx]--;
            dIdx = (dIdx + 1) % 4;
            sIdx = (sIdx + 1) % 2;
        }
        return matrix;
    }
};